### PR TITLE
feat: check for qubit count and presence of native gates at submission to AnyonQPU

### LINKED ISSUE #164: Validate number of qubits before submitting jobs in AnyonQPU

### LINKED ISSUE #166: Validate that gates are native before submitting jobs in AnyonQPU

### PR DESCRIPTION
**Pull request by fSosaRey**
Closes #164, closes #166

### CODE REVIEW

**Review by plerouxanyon — CHANGES_REQUESTED** [review]

File: src/anyon/qpu_interface.jl [hunk 1]
```diff
@@ -339,6 +339,9 @@ get_metadata(qpu::AbstractQPU) =
 get_native_gate_types(qpu::AbstractQPU) = 
     throw(NotImplementedError(:get_native_gate_types,qpu))
 
+get_num_qubits(qpu::AbstractQPU) =
```
[the remark this answers is not in the document]
Comment (fSosaRey) [reply src/anyon/qpu_interface.jl:345]: get_num_qubits(::AnyonQPU) is implemented, so either we throw a NonImplementedError for any other AbstractQPU, or the default MethodError will be thrown

[the remark this answers is not in the document]
Comment (plerouxanyon) [reply src/anyon/qpu_interface.jl:345]: I think we should remove the `get_num_qubits` from the abstract QPU interface. It means that code that depends on an AbstractQPU interface is not guaranteed to work with an AbstractQPU. It breaks the Liskov substitution principle. Allowing partial interfaces prevents us from building robust code on top of AnyonQPU.

File: src/anyon/qpu_interface.jl [hunk 2]
```diff
@@ -384,6 +397,20 @@ function read_response_body(body::Vector{UInt8})::String
     return body_string
 end
 
+function transpile_and_run_job(
+    qpu::AbstractQPU, 
+    circuit::QuantumCircuit,
+    num_repetitions::Integer;
+    transpiler::Transpiler=TrivialTranspiler()
```
Comment (plerouxanyon) [inline src/anyon/qpu_interface.jl:404]: Remove the transpiler argument and use `get_transpiler(qpu)`. This is meant as a user-friendly "run", but without worrying about transpilers. If the user wants to use their own transpiler then they can call transpile themselves.

File: src/anyon/anyon.jl [hunk 3]
```diff
@@ -25,37 +26,105 @@ struct AnyonQPU <: AbstractQPU
...
     "serial_number" =>"ANYK202201",
+    "qubit_count"   =>6,
+    "connectivity_type"  =>line_connectivity_label
 )
 
 get_client(qpu_service::AnyonQPU)=qpu_service.client
 
+get_num_qubits(qpu::AnyonQPU)=get_metadata(qpu)["qubit_count"]
+
+get_connectivity(qpu::AnyonQPU)="1──2──3──4──5──6"
```
Comment (plerouxanyon) [inline src/anyon/anyon.jl:43]: Rename to `print_connectivity(io, qpu)` and draw to `io`.

File: src/anyon/anyon.jl [hunk 4]
```diff
@@ -25,37 +26,105 @@ struct AnyonQPU <: AbstractQPU
...
+            
+        targets=get_connected_qubits(gate)
+
+        return (abs(targets[1]-targets[2])==1)        
+    end
+        
+    return (typeof(gate) in set_of_native_gates)
+end
+
+function is_native_circuit(qpu::AnyonQPU,circuit::QuantumCircuit)::Bool
```
Comment (plerouxanyon) [inline src/anyon/anyon.jl:88]: It is unexpected that `is_native_circuit` throws an error if it returns a bool. I think it is better to have this return bool and throw an error in `run_job` if the circuit is not native.

**Review by plerouxanyon — CHANGES_REQUESTED** [review]

**Comment by fSosaRey** [comment]
Closes #167

File: src/anyon/qpu_interface.jl [hunk 5]
```diff
@@ -336,8 +336,17 @@ abstract type AbstractQPU end
...
 
-get_native_gate_types(qpu::AbstractQPU) = 
-    throw(NotImplementedError(:get_native_gate_types,qpu))
+is_native_gate(qpu::AbstractQPU,::AbstractGate) = 
+    throw(NotImplementedError(:is_native_gate,qpu))
+
+is_native_circuit(qpu::AbstractQPU,::AbstractGate) =
+    throw(NotImplementedError(:is_native_circuit,qpu))
+
+print_connectivity(qpu::AbstractQPU,io::IO=stdout) =
```
Comment (plerouxanyon) [inline src/anyon/qpu_interface.jl:345]: This cannot be implemented for all QPUs. What does the connectivity of a VirtualQPU look like?

**Review by plerouxanyon — CHANGES_REQUESTED** [review]

**Review by plerouxanyon — APPROVED** [review]

**Pull request merged**